### PR TITLE
Map editor crashes - if items assigned to players without spawns

### PR DESCRIPTION
**Pull request by TamasKiss**
- After a player removes a spawn point which is selected, the owner
   remains selected in the DropdownList after trying to spawn something
   the EditorActorBrush gets created with an invalid owner.

 - This fixes this by checking if the removed spawnpoint is the selected
   one, then falls back to the previous entry.

### CODE REVIEW

**Comment by pchote** [comment]
This looks broadly correct, but I haven't had a chance to test this yet, and there will be a couple of minor code style nits to look into.  I should hopefully be able to look at this in a few days if @reaperrr or @abcdefg30 don't beat me to it.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 1]
```diff
@@ -94,6 +86,24 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
 				ownersDropDown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 270, owners, setupItem);
 			};
 
+			editorLayer.OnSpawnRemoved = () =>
+			{
+				if (editorLayer.Players.Players.Values.Any(p => p.Name.Equals(selectedOwner.Name)))
+				{
+					return;
+				}
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: Since it's just a single line, you can drop the curly brackets here.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 2]
```diff
@@ -94,6 +86,24 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
...
+
+				var orderedOwners = editorLayer.Players.Players.Values.ToList();
+				orderedOwners.Add(selectedOwner);
+				orderedOwners = orderedOwners.OrderBy(p => p.Name).ToList();
+
+				var indexofPreviousSelectedOwned = orderedOwners.IndexOf(selectedOwner);
+
+				SelectOwner(indexofPreviousSelectedOwned > 0
+					? orderedOwners.ElementAt(indexofPreviousSelectedOwned - 1)
+					: orderedOwners.FirstOrDefault());
```
Comment (MunWolf) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: This is bad since orderedOwners.FirstOrDefault can still return a null reference if orderedOwners is empty, and SelectOwner assumes that it will get a valid option so it crashes on null.

Comment (MunWolf) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: You could probably fix it by doing a null check in SelectOwner and then setting everything to default values if it isnt valid.
Unless orderedOwners can never be empty in which case ignore my comment.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 3]
```diff
@@ -94,6 +86,24 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
...
+			editorLayer.OnSpawnRemoved = () =>
+			{
+				if (editorLayer.Players.Players.Values.Any(p => p.Name.Equals(selectedOwner.Name)))
+				{
+					return;
+				}
+
+				var orderedOwners = editorLayer.Players.Players.Values.ToList();
+				orderedOwners.Add(selectedOwner);
+				orderedOwners = orderedOwners.OrderBy(p => p.Name).ToList();
```
Comment (MunWolf) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: Any reason why you dont sort inplace with ```orderedOwners.Sort((x, y) => x.Name.CompareTo(y.Name));```?

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 4]
```diff
@@ -72,15 +72,7 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
...
-				var item = ScrollItemWidget.Setup(template, () => selectedOwner == option, () =>
-				{
-					selectedOwner = option;
-
-					ownersDropDown.Text = selectedOwner.Name;
-					ownersDropDown.TextColor = selectedOwner.Color.RGB;
-
-					InitializeActorPreviews();
-				});
+				var item = ScrollItemWidget.Setup(template, () => selectedOwner == option, () => { SelectOwner(option); });
```
Comment (MunWolf) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: ```() => { SelectOwner(option); }```
can be rewritten to
```() => SelectOwner(option)```

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 5]
```diff
@@ -131,7 +141,7 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
 
 				var tooltipText = (tooltip == null ? "Type: " : tooltip.Name + "\nType: ") + a.Name;
 				allActorsTemp.Add(new ActorSelectorActor(a, editorData.Categories, searchTerms.ToArray(), tooltipText));
- 			}
+			}
```
Comment (MunWolf) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: Why is this change here?

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs (hunk 2 above)
Comment (MunWolf) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: Actually I had a look and we assume above that orderedOwners can never be empty so change it from FirstOrDefault to just First and we will call it a day.

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs (hunk 5 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: Looks like he's removing a stray space, which should be fine (imho).

Comment (MunWolf) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: Lets say its fine then :)

**Comment by TamasKiss** [comment]
I've updated the commit!

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 6]
```diff
@@ -94,6 +86,22 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
...
 			};
 
+			editorLayer.OnSpawnRemoved = () =>
+			{
+				if (editorLayer.Players.Players.Values.Any(p => p.Name.Equals(selectedOwner.Name)))
+					return;
+
+				var orderedOwners = editorLayer.Players.Players.Values.ToList();
+				orderedOwners.Add(selectedOwner);
+				orderedOwners.Sort((x, y) => String.Compare(x.Name, y.Name, StringComparison.Ordinal));
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: > OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs:L96: [SA1121] Use the built-in type alias 'string' rather than String or System.String.

Use lower-case _s_tring.Compare.

**Comment by MunWolf** [comment]
@abcdefg30 he still has not fixed one of my comments (FirstOrDefault -> First)
Why did you remove the fixup tag?

**Comment by abcdefg30** [comment]
Ah sorry, I thought he fixed that.

**Comment by TamasKiss** [comment]
It should be all fixed up now!

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 7]
```diff
@@ -94,6 +86,22 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
 				ownersDropDown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 270, owners, setupItem);
 			};
 
+			editorLayer.OnSpawnRemoved = () =>
+			{
+				if (editorLayer.Players.Players.Values.Any(p => p.Name.Equals(selectedOwner.Name)))
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: Style nit: Please prefer `==` over `.Equals()`

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 8]
```diff
@@ -94,6 +86,22 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
...
+					return;
+
+				var orderedOwners = editorLayer.Players.Players.Values.ToList();
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: It should be possible to simplify this down to something like the following (but note that I haven't tested this, so you may need to tweak it):
```csharp
// Fall back to the previous owner in the list.
var previousOwners = editorLayer.Players.Players.Values
	.Append(selectedOwner)
	.OrderBy(x => x.Name)
	.TakeWhile(x => x != selectedOwner);

SelectOwner(previousOwners.LastOrDefault() ?? previousOwners.First());
```

File: OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs [hunk 9]
```diff
@@ -202,6 +210,16 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
 			InitializeActorPreviews();
 		}
 
+		void SelectOwner(PlayerReference option)
+		{
+			selectedOwner = option;
+
+			ownersDropDown.Text = selectedOwner.Name;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: Style nit: please use `option` here instead of `selectedOwner` for consistency.

Comment (MunWolf) [reply OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs]: This was copy pasted over from up top btw, this code already existed. (Still worth changing it)

**Comment by pchote** [comment]
No activity here in quite some time, so closing in favour of #15628.
